src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,370 @@
+					if( extracted.stackSize > remainingSize )
+					{
+						// Something broke. It should never return more than we requested... We're going to silently eat the remainder
+						AELog.warn( "Mod that provided item handler {} is broken. Returned {} items, even though we requested {}.",
+								itemHandler.getClass().getSimpleName(), extracted.stackSize, remainingSize );
+						extracted.stackSize = remainingSize;
+					}
+
+					// We're just gonna use the first stack we get our hands on as the template for the rest
+					if( gathered == null )
+					{
+						gathered = extracted;
+					}
+					else
+					{
+						gathered.stackSize += extracted.stackSize;
+					}
+					remainingSize -= gathered.stackSize;
+				}
+			}
+			while( extracted != null && remainingSize > 0 );
+
+			// Done?
+			if( remainingSize <= 0 )
+			{
+				break;
+			}
+		}
+
+		if( gathered != null )
+		{
+			if( mode == Actionable.MODULATE )
+			{
+				this.onTick();
+			}
+
+			return AEItemStack.create( gathered );
+		}
+
+		return null;
+
+	}
+
+	public TickRateModulation onTick()
+	{
+		LinkedList<IAEItemStack> changes = new LinkedList<IAEItemStack>();
+
+		int slots = itemHandler.getSlots();
+
+		// Make room for new slots
+		if ( slots > cachedStacks.length ) {
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		for( int slot = 0; slot < slots; slot++ )
+		{
+			// Save the old stuff
+			ItemStack oldIS = cachedStacks[slot];
+			IAEItemStack oldAeIS = cachedAeStacks[slot];
+
+			ItemStack newIS = itemHandler.getStackInSlot( slot );
+
+			if( this.isDifferent( newIS, oldIS ) )
+			{
+				// Completely different item
+				cachedStacks[slot] = newIS;
+				cachedAeStacks[slot] = AEItemStack.create( newIS );
+
+				// If we had a stack previously in this slot, notify the newtork about its disappearance
+				if( oldAeIS != null )
+				{
+					oldAeIS.setStackSize( -oldAeIS.getStackSize() );
+					changes.add( oldAeIS );
+				}
+
+				// Notify the network about the new stack. Note that this is null if newIS was null
+				if( cachedAeStacks[slot] != null )
+				{
+					changes.add( cachedAeStacks[slot] );
+				}
+			}
+			else
+			{
+				// Still the same item, but amount might have changed
+				int newSize = ( newIS == null ? 0 : newIS.stackSize );
+				int diff = newSize - ( oldIS == null ? 0 : oldIS.stackSize );
+
+				IAEItemStack stack = ( oldAeIS == null ? AEItemStack.create( newIS ) : oldAeIS.copy() );
+				if( stack != null )
+				{
+					stack.setStackSize( newSize );
+				}
+
+				if( diff != 0 && stack != null )
+				{
+					cachedStacks[slot] = newIS;
+					cachedAeStacks[slot] = stack;
+
+					final IAEItemStack a = stack.copy();
+					a.setStackSize( diff );
+					changes.add( a );
+				}
+			}
+		}
+
+		// Handle cases where the number of slots actually is lower now than before
+		if( slots < cachedStacks.length )
+		{
+			for (int slot = slots; slot < cachedStacks.length; slot++ )
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) '(' is not followed by whitespace. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)